### PR TITLE
Upgrade Logger

### PR DESCRIPTION
**Pull request by nholland94**
Here's the first part of the Logger RFC changes #1798. This replaces the implementation and interface of the `Logger` module. I went ahead and updated many of the log messages to use the new json metadata system instead of logging raw sexps, but some of the nastier types have been skipped and will be done in future PRs. In the short term, this will make our logs much uglier, so we will want to have a discussion regarding whether we want to merge this or not before the new log processor lands.

Here is an pretty-printed example of what a log message looks like with the new format:

```json
{
  "timestamp": "2019-03-09 00:18:02.389201Z",
  "level": "Info",
  "source": {
    "module": "Coda_main",
    "location": "File \"src/app/cli/src/coda_main.ml\", line 1176, characters 18-25"
  },
  "message": "Added payment $user_command to pool successfully",
  "metadata": {
    "host": "127.0.0.1",
    "port": 23000,
    "user_command": {
      "payload": {
        "common": {
          "fee": 0,
          "nonce": 2,
          "memo": "aSF6MHmQgJThESHQQjVKfB9VtkgsoaUeGyUN/R7Q7vk="
        },
        "body": [
          "Payment",
          {
            "receiver": "KF/gCh9fJ7rsTA72qygPI+BW1hLOGmNHIj+jXsv30WJpIkrVxzsBAAAA",
            "amount": 10
          }
        ]
      },
      "sender": "ASiJrrItmZLlhj1zdGKDojmEcCFS06o/dYdZWAm6LBAQ82uwBLZoAQAAAQ==",
      "signature": "KN2IETEDyZBRtyMkK5zQH2PYDVh/hsNbQ58fJoxQ37ZNP0e9R9QBAAAorxfG6FpvRtShsc6o9xARAMSagrxl/nmYRtRwbT4rpcAAAAAAAAAAAA=="
    }
  }
}
```

### CODE REVIEW

**Comment by jkrauska** [comment]
Question: I'm hoping that logs stay single line. 
The example you gave is multiline (after a jq parse? I hope?)

Can you also please maintain host and pid?
Otherwise you're taking away information (breaking compatibility) from the existing solution.

**Comment by nholland94** [comment]
@jkrauska The example I gave is only pretty printed for readability. The raw logs are still single line. I can add host and port back in if you need that, though it doesn't really make sense to me to put that in every single log statement. Can I put it behind a cli flag to enable it?

**Comment by cmr** [comment]
My argument for including PID by default (not host) is that, in integration tests, we're ~always looking at intermingled log streams. Hopefully most of our debugging and thus log delving happens with integration-or-smaller tests.

File: src/app/logproc/filter.ml [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+  let pad b p = b *> p <* b
+  let wrap (l, r) p = l *> p <* r
+  let maybe p = p >>| Option.some <|> return None
+
+  (* string w/ commit on first char *)
+  let stringc s =
+    assert (String.length s > 0);
+    let h = String.get s 0 in
+    let t = String.sub s ~pos:1 ~len:(String.length s - 1) in
+    lift2 (fun h' t' -> print_endline "fuck"; String.of_char h' ^ t') (char h <* commit) (string t)
```
Comment (cmr) [inline src/app/logproc/filter.ml]: leftover debug

Comment (nholland94) [reply src/app/logproc/filter.ml]: Yes, one I will squash from history as well :)

**Comment by nholland94** [comment]
> My argument for including PID by default (not host) is that, in integration tests, we're ~always looking at intermingled log streams. Hopefully most of our debugging and thus log delving happens with integration-or-smaller tests.

I find this argument for pid compelling so I have added it back in. Waiting for further discussion before touching host.

**Comment by nholland94** [comment]
I have finished making the first round of requested changes. Here are some demonstrations of filters you can create now:

find all warnings and errors: `logproc -f '.level in ["Warn", "Error", "Fatal", "Faulty_peer"]`
find all logs with payments: `logproc -f '.metadata.user_command.payload.body[0] == "Payment"'`
using a regex on a string field: `logproc -f '.level match /^Fa.{1,2}l/`

File: scripts/test.py [hunk 2]
```diff
@@ -147,17 +154,17 @@ def make_log_dir():
         print('- %s:' % profile)
         build_log = os.path.join(profile_dir, 'build.log')
         run_cmd(
-            'dune build --display=progress --profile=%s %s 2> %s'
-                % (profile, coda_exe_path, build_log),
+            'dune build --display=progress --profile=%s %s %s 2> %s'
```
Comment (cmr) [inline scripts/test.py]: if `logproc` is already in the `PATH`, can we use that instead of building a copy?

**Pull request merged**